File: cpp/450_DeleteNodeInBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root)
            return root;
        
        if (root->val > key) {
            root->left = deleteNode(root->left, key);
        } else if (root->val < key) {
            root->right = deleteNode(root->right, key);
        } else {
            if (root->left == NULL && root->right == NULL) {
                root = NULL;
            } else if (root->right != NULL) {
                root->val = successor(root);
                root->right = deleteNode(root->right, root->val);
            } else {
                root->val = predecessor(root);
                root->left = deleteNode(root->left, root->val);
            }
        }
        return root;
    }
    
    int successor(TreeNode* root) {
        root = root->right;
        while (root->left)
            root = root->left;
        return root->val;
    }
    
    int predecessor(TreeNode* root) {
        root = root->left;
        while (root->right)
            root = root->right;
        return root->val;
    }
};

File: swift/450_DeleteNodeInBST.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func deleteNode(_ node: TreeNode?, _ key: Int) -> TreeNode? {
        var root = node
        if root == nil { return nil }
        
        if key < root!.val {
            root?.left = deleteNode(root?.left, key)
        } else if key > root!.val {
            root?.right = deleteNode(root?.right, key)
        } else {
            if root?.right == nil && root?.left == nil {
                root = nil
            } else if root?.right != nil {
                root!.val = successor(root)
                root?.right = deleteNode(root?.right, root!.val)
            } else {
                root!.val = predecessor(root)
                root?.left = deleteNode(root?.left, root!.val)
            }
        }
        return root
    }
    
    func successor(_ root: TreeNode?) -> Int {
        var node = root
        node = node?.right
        while node?.left != nil {
            node = node?.left
        }
        return node!.val
    }
    
    func predecessor(_ root: TreeNode?) -> Int {
        var node = root
        node = node?.left
        while node?.right != nil {
            node = node?.right
        }
        return node!.val
    }
}
